### PR TITLE
Moves "Patterns" command to site editor main navigation

### PR DESCRIPTION
**Pull request by bacoords**
## What?

As a follow up to [this comment](https://github.com/WordPress/gutenberg/pull/61287#pullrequestreview-2041073623), this PR moves the "Patterns" Command from `admin-navigation-commands` to `site-editor-navigation-commands`.

## Why?

The cleans up the code needed for the admin commands file as well as provides consistency in the navigation experience. Now the initial list of commands exactly matches the global navigation in the site editor. 

It's worth noting that this removes the command from non-administrator users, even if they could technically access `edit.php?post_type=wp_block` and be directed there. 

That said there are a number of commands that should be added for non-admin users that could send them to the non-site editor screens and that's really outside the scope of this PR.

## Testing Instructions
1. Open the site editor.
2. Click cmd-k to trigger the Command Palette
3. Start searching for Patterns
4. Select Patterns and confirm you're navigated to the correct screen

## Screenshots or screencast 

Screenshot 2024-05-06 at 3 21 56 PM

### CODE REVIEW

**Review by t-hamano — COMMENTED** [review]
Thanks for the PR!

I think the original Patterns command was available to all themes and to all users. With this PR, if you [cannot access the site editor or are not using a block theme](https://github.com/WordPress/gutenberg/blob/6769a2ff93aec9feaf3d9bcada94cce9e1adef07/packages/core-commands/src/site-editor-navigation-commands.js#L265-L267), you will not be able to use the Patterns commands.

My understanding is that this PR is a refactoring and I think the previous behavior should be fully preserved.

In my opinion, we need an approach like below.

```javascript
function useSiteEditorBasicNavigationCommands() {
	const commands = useMemo( () => {
		const result = [];

		if ( isTemplatesAccessible && isBlockBasedTheme ) {
			// Command for when the current theme is a block theme and the user has access to the site editor
			Navigation, Styles, Pages, Templates
		}

		result.push( {
			...
			callback: ( { close } ) => {
				// First, check if the current user has access to the site editor.
				if ( isTemplatesAccessible ) {
					const args = {
						path: '/patterns',
					};
					if ( isSiteEditor ) {
						// If you are already in the site editor
						history.push( args );
					} else {
						// If you are outside the site editor
						document.location = addQueryArgs( 'site-editor.php', args );
					}
					close();
				} else {
					// If a user cannot access the site editor
					document.location.href = 'edit.php?post_type=wp_block';
				}
			},
		} );

		return result;
	}, [] );

	return {
		commands,
		isLoading: false,
	};
}
```

File: packages/core-commands/src/site-editor-navigation-commands.js [hunk 1]
```diff
@@ -338,6 +339,24 @@ function useSiteEditorBasicNavigationCommands() {
 			},
 		} );
 
+		result.push( {
+			name: 'core/manage-reusable-blocks',
```
Comment (t-hamano) [inline packages/core-commands/src/site-editor-navigation-commands.js]:
```suggestion
			name: 'core/edit-site/open-patterns',
```

There are no hard and fast rules regarding this property, but it is a good idea to make it consistent with other commands.

**Comment by bacoords** [comment]
Thanks for the feedback @t-hamano ! I've updated the PR but the main issue I've had is that `.canUser( 'read', 'templates' )` was returning true for "Editor" level users in the block editor so they were sent to the Site Editor, (perhaps because they're able to "read" and pull in the templates for the template switching and preview functionality)? 

I added an additional hook based on `.canUser( 'create', 'templates' )` to better determine if the user is able to open the site editor or should be sent to the original admin screen for Patterns, but would love some testing on this.

**Comment by t-hamano** [comment]
> the main issue I've had is that `.canUser( 'read', 'templates' )` was returning true for "Editor" level users in the block editor so they were sent to the Site Editor

After researching this issue, I believe that returning true is an unintended regression. See #61419.

I think we need to fix this permission issue before this PR and #61287.

File: packages/core-commands/src/site-editor-navigation-commands.js [hunk 2]
```diff
@@ -258,83 +263,105 @@ function useSiteEditorBasicNavigationCommands() {
 		'site-editor.php'
 	);
 	const isTemplatesAccessible = useIsTemplatesAccessible();
+	const isTemplatesEditable = useIsTemplatesEditable();
```
Comment (t-hamano) [inline packages/core-commands/src/site-editor-navigation-commands.js]: Based on [this discussion](https://github.com/WordPress/gutenberg/pull/61444#discussion_r1592841651), it seems better not to create custom hooks. You will only need to check whether it is possible to create a template, as shown below.

```javascript
const canCreateTemplate = useSelect( ( select ) => {
	return select( coreStore ).canUser( 'create', 'templates' );
}, [] );
```

**Review by t-hamano — COMMENTED** [review]
An issue with user permissions was fixed in #61444. Could you rebase this branch with the latest trunk?

File: packages/core-commands/src/site-editor-navigation-commands.js (hunk 2 above)
Comment (bacoords) [reply packages/core-commands/src/site-editor-navigation-commands.js]: Thanks - this makes sense and uses the same `canCreateTemplate` value to determine whether the user should be sent to the classic admin screen for patterns or to the site editor's Pattern screen.

**Comment by bacoords** [comment]
> Could you rebase this branch with the latest trunk?

This has been rebased.

**Review by t-hamano — APPROVED** [review]
LGTM! All commands are working as before.


### Admin, Block theme

- Navigation, Styles, Pages, Templates: displayed
- Patterns: `wp-admin/site-editor.php?path=/patterns`

### Editor, Block theme

- Navigation, Styles, Pages, Templates: Not displayed
- Pattern: > `wp-admin/edit.php?post_type=wp_block`

### Admin, Classic theme

- Navigation, Styles, Pages, Templates: Not displayed
- Pattern: > `wp-admin/site-editor.php?path=/patterns`

### Editor, Classic theme

- Navigation, Styles, Pages, Templates: Not displayed
- Pattern: > `wp-admin/edit.php?post_type=wp_block`

**Pull request merged**